device/core/src/x509_authentication_provider.ts

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import { AuthenticationProvider, AuthenticationType, errors, DeviceCredentials, X509 } from 'azure-iot-common';
+
+export class X509AuthenticationProvider implements AuthenticationProvider {
+  type: AuthenticationType = AuthenticationType.X509;
+  private _credentials: DeviceCredentials;
+
+  constructor(credentials: DeviceCredentials) {
+    /*Codes_SRS_NODE_X509_AUTHENTICATION_PROVIDER_16_001: [The `constructor` shall store the credentials passed as argument.]*/
+    this._credentials = credentials;
+  }
+
+  getDeviceCredentials(callback: (err: Error, credentials?: DeviceCredentials) => void): void {
+    /*Codes_SRS_NODE_X509_AUTHENTICATION_PROVIDER_16_002: [The `getDeviceCredentials` method shall call its callback with a `null` error object and the stored device credentials as a second argument.]*/
+    callback(null, this._credentials);
+  }
+
+  setX509Options(x509: X509): void {
+    /*Codes_SRS_NODE_X509_AUTHENTICATION_PROVIDER_16_003: [The `setX509Options` method shall store the `X509` object passed as an argument with the existing credentials.]*/
+    this._credentials.x509 = x509;
+  }
```

Comment:
Is there a use case for this function?  If not, suggest removing. The factory requires a cert, why would you ever change it?

Reply:
the problem comes with the Client code, where it used to be that the user would create a client with an x509 connection string (not including the cert/key) and then call setOptions to add the certificate. since we can't wait until setOptions to create an AuthenticationProvider to give to the transport, we have to have a thing that calls the AuthenticationProvider from the body of setOptions to add the cert/key. it's mostly to be compatible with legacy code.

---
In reply to: [157025104](https://github.com/Azure/azure-iot-sdk-node/pull/179#discussion_r157025104) [](ancestors = 157025104)

Reply:
got it.  that make sense.  please mark as by design.

---
In reply to: [157032885](https://github.com/Azure/azure-iot-sdk-node/pull/179#discussion_r157032885) [](ancestors = 157032885,157025104)